Userland/Services/WindowServer/WindowFrame.h

```diff
@@ -20,21 +20,46 @@ class Button;
 class Menu;
 class MouseEvent;
 class Window;
+class Screen;
 
 class WindowFrame {
 public:
+    class RenderedCache {
+        friend class WindowFrame;
+
+    public:
+        void paint(WindowFrame&, Gfx::Painter&, const Gfx::IntRect&);
+        void render(WindowFrame&, Screen&);
+        Optional<HitTestResult> hit_test(WindowFrame&, Gfx::IntPoint const&, Gfx::IntPoint const&);
+
+    private:
+        RefPtr<Gfx::Bitmap> m_top_bottom;
+        RefPtr<Gfx::Bitmap> m_left_right;
+        int m_bottom_y { 0 }; // y-offset in m_top_bottom for the bottom half
+        int m_right_x { 0 };  // x-offset in m_left_right for the right half
+        bool m_shadow_dirty { true };
+        bool m_dirty { true };
+    };
+    friend class RenderedCache;
+
     static void reload_config();
 
     explicit WindowFrame(Window&);
     ~WindowFrame();
 
+    void window_was_constructed();
```

Comment:
Can this take a `Badge<Window>`, since it's supposed to be called there ? (Not sure if it would work given that Window is forward-declared, but if it's possible it would be a good thing).